File: backend/src/models/book.js
const mongoose = require("mongoose");
const Schema = mongoose.Schema;

const bookSchema = new Schema(
  {
    title: {
      type: String,
      required: true,
    },
    author: {
      type: String,
      required: true,
    },
    description: {
      type: String,
      required: true,
    },
    genre: {
      type: String,
      required: true,
    },
    pages: {
      type: Number,
      required: true,
    },
    coverImage: {
      type: String,
      required: true,
    },
    pdf: {
      fileName: {
        type: String,
        required: true,
      },
      data: {
        type: String,
        required: true,
      },
    },
    ratings: {
      averageRating: {
        type: Number,
        min: 0,
        max: 5,
      },
      numberOfRatings: {
        type: Number,
        default: 0,
      },
    },
    creator: {
      type: Schema.Types.ObjectId,
      ref: "User",
      required: true,
    },
    reviews: [
      {
        type: Schema.Types.ObjectId,
        ref: "Review",
      },
    ],
  },
  { timestamps: true }
);

const Book = mongoose.model("Book", bookSchema);
module.exports = Book;

File: frontend/lib/features/home/data/models/book.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'book.freezed.dart';
part 'book.g.dart';

@freezed
class Book with _$Book {
  const factory Book({
    required String title,
    required String author,
    required String description,
    required String genre,
    required int pages,
    required CoverImage coverImage,
    required Pdf pdf,
    required Ratings ratings,
    required List<String> reviews,
  }) = _Book;

  factory Book.fromJson(Map<String, dynamic> json) => _$BookFromJson(json);
}

@freezed
class CoverImage with _$CoverImage {
  const factory CoverImage({
    required List<int> data,
    required String contentType,
  }) = _CoverImage;

  factory CoverImage.fromJson(Map<String, dynamic> json) =>
      _$CoverImageFromJson(json);
}

@freezed
class Pdf with _$Pdf {
  const factory Pdf({
    required String filename,
    required List<int> data,
    required String contentType,
  }) = _Pdf;

  factory Pdf.fromJson(Map<String, dynamic> json) => _$PdfFromJson(json);
}

@freezed
class Ratings with _$Ratings {
  const factory Ratings({
    required double averageRating,
    required int numberOfRatings,
  }) = _Ratings;

  factory Ratings.fromJson(Map<String, dynamic> json) =>
      _$RatingsFromJson(json);
}
